pkg/operator/controller/ingress/deployment.go

```diff
@@ -413,7 +413,7 @@ func desiredRouterDeployment(ci *operatorv1.IngressController, ingressController
 	})
 
 	if len(ci.Status.Domain) > 0 {
-		env = append(env, corev1.EnvVar{Name: "ROUTER_CANONICAL_HOSTNAME", Value: ci.Status.Domain})
+		env = append(env, corev1.EnvVar{Name: "ROUTER_CANONICAL_HOSTNAME", Value: "router-" + ci.Name + "." + ci.Status.Domain})
```

Comment:
Do we have a test for shards? Will this work for shards?

Reply:
I tested this  and it is working for shards->

1. [miheer@localhost cluster-ingress-operator]$ cat router-internal.yaml 
apiVersion: v1
items:
- apiVersion: operator.openshift.io/v1
  kind: IngressController
  metadata:
    name: sharded
    namespace: openshift-ingress-operator
  spec:
    domain: apps1.ci-ln-q3j9m9k-f76d1.origin-ci-int-gce.dev.openshift.com
    nodePlacement:
      nodeSelector:
        matchLabels:
          node-role.kubernetes.io/worker: ""
    routeSelector:
      matchLabels:
        type: sharded
  status: {}
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
[miheer@localhost cluster-ingress-operator]$



2. [miheer@localhost cluster-ingress-operator]$ oc apply -f router-internal.yaml 


3. oc edit route and add the label  type: sharded

[miheer@localhost cluster-ingress-operator]$ oc edit route django-psql-example
route.route.openshift.io/django-psql-example edited

4. You will this route getting admitted in the new router for sharding as follows ->
[miheer@localhost cluster-ingress-operator]$ oc get route -o yaml
apiVersion: v1
items:
- apiVersion: route.openshift.io/v1
  kind: Route
  metadata:
    annotations:
      openshift.io/generated-by: OpenShiftNewApp
      openshift.io/host.generated: "true"
    creationTimestamp: "2021-05-13T12:28:38Z"
    labels:
      app: django-psql-example
      template: django-psql-example
      type: sharded
    managedFields:
    - apiVersion: route.openshift.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:openshift.io/generated-by: {}
          f:labels:
            .: {}
            f:app: {}
            f:template: {}
        f:spec:
          f:to:
            f:kind: {}
            f:name: {}
            f:weight: {}
          f:wildcardPolicy: {}
      manager: oc
      operation: Update
      time: "2021-05-13T12:28:38Z"
    - apiVersion: route.openshift.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:ingress: {}
      manager: openshift-router
      operation: Update
      time: "2021-05-13T12:28:38Z"
    - apiVersion: route.openshift.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:labels:
            f:type: {}
      manager: kubectl-edit
      operation: Update
      time: "2021-05-13T12:55:10Z"
    name: django-psql-example
    namespace: test-ingress
    resourceVersion: "44555"
    uid: 097ffefb-9f5a-4633-bd9b-f6f90464a4a0
  spec:
    host: django-psql-example-test-ingress.apps.ci-ln-q3j9m9k-f76d1.origin-ci-int-gce.dev.openshift.com
    to:
      kind: Service
      name: django-psql-example
      weight: 100
    wildcardPolicy: None
  status:
    ingress:
    - conditions:
      - lastTransitionTime: "2021-05-13T12:28:38Z"
        status: "True"
        type: Admitted
      host: django-psql-example-test-ingress.apps.ci-ln-q3j9m9k-f76d1.origin-ci-int-gce.dev.openshift.com
      routerCanonicalHostname: router-default.apps.ci-ln-q3j9m9k-f76d1.origin-ci-int-gce.dev.openshift.com
      routerName: default
      wildcardPolicy: None
    - conditions:
      - lastTransitionTime: "2021-05-13T12:55:10Z"
        status: "True"
        type: Admitted
      host: django-psql-example-test-ingress.apps.ci-ln-q3j9m9k-f76d1.origin-ci-int-gce.dev.openshift.com
      routerCanonicalHostname: router-sharded.apps1.ci-ln-q3j9m9k-f76d1.origin-ci-int-gce.dev.openshift.com
      routerName: sharded
      wildcardPolicy: None
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
[miheer@localhost cluster-ingress-operator]$